tests/test_arbitrage_integration.py

```diff
@@ -23,4 +23,4 @@ def test_arbitrage_real_block():
     assert (
         arbitrage.profit_token_address == "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"
     )
-    assert arbitrage.profit_amount == 53560707941943273628
```

Comment:
There's actually 2 arbitrages in this block!

The one I [referenced](https://etherscan.io/tx/0x448245bf1a507b73516c4eeee01611927dada6610bf26d403012f2e66800d8f0). And the [https://etherscan.io/tx/0xfcf4558f6432689ea57737fe63124a5ec39fd6ba6aaf198df13a825dd599bffc]. I have to admit I have no idea what is going on in 0xfcf4...

Here's the full list of swaps
```
[Swap(abi_name='UniswapV3Pool', transaction_hash='0xfcf4558f6432689ea57737fe63124a5ec39fd6ba6aaf198df13a825dd599bffc', block_number=12914944, trace_address=[0, 0], pool_address='0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640', from_address='0x7ec0b75a98997c927ace6d87958147a105147ea0', to_address='0x0000fee6275dab194ab538a01dd8b18b02b20000', token_in_address='0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', token_in_amount=283588902010, token_out_address='0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', token_out_amount=123848351154563483804, protocol=None, error=None), Swap(abi_name='UniswapV3Pool', transaction_hash='0xfcf4558f6432689ea57737fe63124a5ec39fd6ba6aaf198df13a825dd599bffc', block_number=12914944, trace_address=[0, 0, 2, 0], pool_address='0x7ec0b75a98997c927ace6d87958147a105147ea0', from_address='0x99b42f2b49c395d2a77d973f6009abb5d67da343', to_address='0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640', token_in_address='0x25f8087ead173b73d6e8b84329989a8eea16cf73', token_in_amount=197386080350317054519429, token_out_address='0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', token_out_amount=283588902010, protocol=None, error=None), Swap(abi_name='UniswapV2Pair', transaction_hash='0xfcf4558f6432689ea57737fe63124a5ec39fd6ba6aaf198df13a825dd599bffc', block_number=12914944, trace_address=[0, 0, 2, 0, 2, 1], pool_address='0x99b42f2b49c395d2a77d973f6009abb5d67da343', from_address='0x0000fee6275dab194ab538a01dd8b18b02b20000', to_address='0x7ec0b75a98997c927ace6d87958147a105147ea0', token_in_address='0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', token_in_amount=70287643212620210176, token_out_address='0x25f8087ead173b73d6e8b84329989a8eea16cf73', token_out_amount=197386080350317054519429, protocol=None, error=None)]
```

Simplified and ordered txs:
```
- Swap ([0, 0]): 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48 -> 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
- Swap ([0, 0, 2, 0]): 0x25f8087ead173b73d6e8b84329989a8eea16cf73 -> 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48
- Swap ([0, 0, 2, 0, 2, 1]): 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 -> 0x25f8087ead173b73d6e8b84329989a8eea16cf73
```

Note that this is reversed based on what we expected from etherscan.
Either trace sort order is wrong (root calls come after child calls) or this contract is doing something very funky. I'll continue to investigate but any thoughts would be awesome!

Reply:
Appreciate you digging in and these PRs in general!

For 0xfcf45, I think the trace order is correct and is kind of uncovering an issue with the logic in the comment you left

Arb swaps can happen "out of order" as long as there's a rebalancing of the pool by the end of the transaction

Reply:
Happy to!

I'm going to rework this to work like the old logic but not be dependent on `swap.pool_address`. May work with a combination of `swap.pool_address` and `swap.from_address`.

Reply:
I'm seeing two ways to do this
1. Add 2x Transfer objects to each Swap object (as in mev-inspect-rs where a Swap is simply composed of two transfers, resulting in all 4 relevant addresses, rather than 3)
2. Look for cycles of `swap -> ... -> swap` where 
    - `swap.token_in` matches the preceeding `swap.token_out`
    - `swap[0].token_in == swap[-1].token_out`

(1.) is guaranteed to work, but (2.) may be sufficient. The reasoning for why we can't build perfect circuits can be seen in the example txs below. 0x44824 has the to and from addresses as the MEV bot's contract address.

Here's some clearer summaries on the two different txs in this test block that both need to be picked up:
```
ordered swap set for 0x448245bf1a507b73516c4eeee01611927dada6610bf26d403012f2e66800d8f0
Swap: 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 -> 0x25f8087ead173b73d6e8b84329989a8eea16cf73
Pool: 0x99b42f2b49c395d2a77d973f6009abb5d67da343 | from: 0x5e6732d6416d3f3a6574b61ef1c81316cf8d9287 | to: 0x5e6732d6416d3f3a6574b61ef1c81316cf8d9287

Swap: 0x25f8087ead173b73d6e8b84329989a8eea16cf73 -> 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48
Pool: 0xb9c31afc2df2cd49c2a685ff378c21eb7d115541 | from: 0x5e6732d6416d3f3a6574b61ef1c81316cf8d9287 | to: 0x5e6732d6416d3f3a6574b61ef1c81316cf8d9287

Swap: 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48 -> 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
Pool: 0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640 | from: 0x5e6732d6416d3f3a6574b61ef1c81316cf8d9287 | to: 0x5e6732d6416d3f3a6574b61ef1c81316cf8d9287
```

```
ordered swap set for 0xfcf4558f6432689ea57737fe63124a5ec39fd6ba6aaf198df13a825dd599bffc
Swap: 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48 -> 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
Pool: 0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640 | from: 0x7ec0b75a98997c927ace6d87958147a105147ea0 | to: 0x0000fee6275dab194ab538a01dd8b18b02b20000

Swap: 0x25f8087ead173b73d6e8b84329989a8eea16cf73 -> 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48
Pool: 0x7ec0b75a98997c927ace6d87958147a105147ea0 | from: 0x99b42f2b49c395d2a77d973f6009abb5d67da343 | to: 0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640

Swap: 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 -> 0x25f8087ead173b73d6e8b84329989a8eea16cf73
Pool: 0x99b42f2b49c395d2a77d973f6009abb5d67da343 | from: 0x0000fee6275dab194ab538a01dd8b18b02b20000 | to: 0x7ec0b75a98997c927ace6d87958147a105147ea0
```

Reply:
it looks like in the first one it's printing the same address for to and from

what's the difference between options 1 and 2 in practice? I'd vote 2 since it stays derivative of the swap

the from_address and to_address for the swap should be the address putting tokens in and the one where the tokens are coming out, so it should be the same as looking at the transfer level. Right now we're indifferent whether those addresses are a contract address or an EOA (better to be able to see where people are keeping tokens)

Reply:
Ok, I've reworked this substantially to implement option 2. Updating the PR description to explain the new algorithm and current short-falls.

Basically we have two types of different arbitrages we need to deal with

Uni router style: MEV_BOT -> Pool A -> Pool B -> MEV_BOT
Standard style: MEV_BOT -> Pool A -> MEV_BOT -> Pool B -> MEV_BOT

The new detection strategy doesn't depend entirely depend on these transfer addresses so it handles both. 

(I'll try to outline this thoroughly in the PR description)

Reply:
Details in PR description!